File: Java/Graph/Floyd Warshall.java
static int n, m, q;
static long INF = Long.MAX_VALUE/10;
public static void main(String[] args) {
    // O(V^3)
    // Floyd Warshall
    n = nextInt();
    m = nextInt();
    q = nextInt();
    long g[][] = new long[n][n];
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            g[i][j] = g[j][i] = INF;
        }
    }
    int a, b;
    long w;
    for (int i = 0; i < m; i++) {
        a = nextInt();
        b = nextInt();
        w = nextLong();
        a--; b--;
        g[a][b] = g[b][a] = min(g[a][b],w);
    }
    for(int k = 0; k < n ; k++){
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                g[i][j] = min(g[i][j], g[i][k] + g[k][j]);
            }
        }
    }
    for (int i = 0; i < q; i++) {
        a = nextInt();
        b = nextInt();
        a--; b--;
        sa.println(g[a][b] == INF? -1 : g[a][b]);
    }
    sa.close();
}

File: cpp/Graph/Floyd Warshall.cpp
const ll oo = 1e18 / 10;
int main(){
    int n, m, q;
    cin >> n >> m >> q;
    vector<vll> g(n, vll(n));
    forn(i, n){
        forab(j, i + 1, n){
            g[i][j] = g[j][i] = oo;
        }
    }
    int a, b;
    ll w;
    forn(i, m){
        cin >> a >> b >> w;
        a--;b--;
        g[a][b] = g[b][a] = min(g[a][b], w);
    }
    //Floyd Warshall O(n^3)
    forn(k, n){
        forn (i, n) {
            forn (j, n) {
                g[i][j] = min(g[i][j], g[i][k] + g[k][j]);
            }
        }
    }
    while(q--){
        cin >> a >> b;
        a--;b--;
        cout << (g[a][b] == oo ? -1: g[a][b]) <<ln;
    }
    return 0;
}